src/components/dynamic-room-size.ts

```diff
@@ -0,0 +1,141 @@
+import { EarlyReplyMode, TextBasedCommandBuilder } from "../command-abstractions/text-based-command-builder.js";
+import { TextBasedCommand } from "../command-abstractions/text-based-command.js";
+
+type voice_limit_entry = {
+    channel: string;
+    new_limit: number;
+    old_limit: number;
+    issuer: string;
+};
+
+export default class DynamicRoomSize extends BotComponent {
+    private database = this.wheatley.database.create_proxy<{
+        voice_limits: voice_limit_entry;
+    }>();
+
+    override async setup(commands: CommandSetBuilder) {
+        commands.add(
+            new TextBasedCommandBuilder("voice-limit", EarlyReplyMode.ephemeral)
+                .set_category("Admin utilities")
+                .set_permissions(Discord.PermissionFlagsBits.MuteMembers)
+                .set_description("Manage voice channel user limits")
+                .add_subcommand(
+                    new TextBasedCommandBuilder("push", EarlyReplyMode.ephemeral)
+                        .set_description("Push new channel limit")
+                        .set_slash(true)
+                        .add_number_option({
+                            title: "limit",
+                            description: "new channel limit",
+                            required: true,
+                        })
+                        .set_handler(this.handle_push.bind(this)),
+                )
+                .add_subcommand(
+                    new TextBasedCommandBuilder("pop", EarlyReplyMode.ephemeral)
+                        .set_description("Pop channel limit")
+                        .set_slash(true)
+                        .set_handler(this.handle_pop.bind(this)),
+                ),
+        );
+    }
+
+    private async handle_push(command: TextBasedCommand, limit: number) {
+        try {
+            if ((await command.get_member()).voice.channelId != command.channel_id) {
+                throw Error("command can only be used in a voice channel you're currently in");
+            }
+            if (!command.channel?.isVoiceBased()) {
+                throw Error("command can only be used in voice channel");
+            }
+            const res = await this.database.voice_limits.findOneAndUpdate(
+                { channel: command.channel_id },
+                {
+                    $set: {
+                        new_limit: limit,
+                        issuer: command.user.id,
+                    },
+                    $setOnInsert: {
+                        channel: command.channel_id,
+                        old_limit: command.channel.userLimit,
+                    },
+                },
+                { upsert: true, returnDocument: "after" },
+            );
+            assert(res);
+            M.log(`🔊 ${command.channel.name} voice limit bumped to ${limit} by ${command.user.displayName}`);
+            await command.channel.setUserLimit(res.new_limit);
```

Comment:
What if `command.channel != (await command.get_member()).voice.channelId`? This might be surprising

Reply:
Ah, I see